tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,384 @@
+
+WriteAtMost(data) ==
+    LET numWriteableWithoutSeeking == Min(ArrayLen(data), lo + BuffSz - curr) IN
+    LET buffOff == curr - lo IN
+    /\ Inv2
+    /\ curr + numWriteableWithoutSeeking <= MaxOffset
+    /\ buff' = ArrayConcat(ArrayConcat(
+            ArraySlice(buff, 0, buffOff),
+            ArraySlice(data, 0, numWriteableWithoutSeeking)),
+            ArraySlice(buff, buffOff + numWriteableWithoutSeeking, ArrayLen(buff)))
+    /\ dirty' = TRUE
+    /\ curr' = curr + numWriteableWithoutSeeking
+    /\ length' = Max(length, curr')
+    /\ UNCHANGED <<lo, diskPos, file_content, file_pointer>>
+
+Next ==
+    \/ FlushBuffer
+    \/ \E offset \in Offset:
+        \/ Seek(offset)
+        \/ SetLength(offset)
+    \/ \E symbol \in Symbols:
+        \/ Read1(symbol)
+        \/ Write1(symbol)
+    \/ \E len \in 1..MaxOffset: \E data \in Array(Symbols, len):
+        \/ WriteAtMost(data)
+        \/ Read(data)
+
+Spec == Init /\ [][Next]_vars
+
+-------------------------------------------------------------------------------
+\* Refinement of general RandomAccessFile
+
+RAF == INSTANCE RandomAccessFile WITH
+    file_content <- LogicalFileContent,
+    file_pointer <- curr
+
+Safety == RAF!Spec
+
+\* Ensure that the various actions behave according to their abstract specifications.
+FlushBufferCorrect == [][FlushBuffer => UNCHANGED RAF!vars]_vars
+SeekCorrect == [][\A offset \in Offset: Seek(offset) => RAF!Seek(offset)]_vars
+SetLengthCorrect == [][\A offset \in Offset: SetLength(offset) => RAF!SetLength(offset)]_vars
+SeekEstablishesInv2 == [][\A offset \in Offset: Seek(offset) => Inv2']_vars
+Write1Correct == [][\A symbol \in Symbols: Write1(symbol) => RAF!Write(SeqToArray(<<symbol>>))]_vars
+Read1Correct == [][\A symbol \in Symbols: Read1(symbol) => RAF!Read(SeqToArray(<<symbol>>))]_vars
+WriteAtMostCorrect == [][\A len \in 1..MaxOffset: \A data \in Array(Symbols, len): WriteAtMost(data) => \E written \in 1..len: RAF!Write(ArraySlice(data, 0, written))]_vars
+ReadCorrect == [][\A len \in 1..MaxOffset: \A data \in Array(Symbols, len): Read(data) => RAF!Read(data)]_vars
+
+-------------------------------------------------------------------------------
+\* Model checking helper definitions
+
+Symmetry == Permutations(Symbols)
+
+Alias == [
+    \* constants
+    BuffSz            |-> BuffSz,
+    MaxOffset         |-> MaxOffset,
+
+    \* regular vars
+    dirty             |-> dirty,
+    length            |-> length,
+    curr              |-> curr,
+    lo                |-> lo,
+    buff              |-> buff,
+    diskPos           |-> diskPos,
+    file_content      |-> file_content,
+    file_pointer      |-> file_pointer,
+
+    \* abstract vars
+    abstract_contents |-> LogicalFileContent]
+
+===============================================================================
+
+--------------------------- MODULE RandomAccessFile ---------------------------
```

Comment:
I think it's still worth a brief documentation string here, as it's the place I jumped to first to ensure I understood what was going on.  For many folks, even interacting with files in this manner (rather than wholesale reads or writes) is still quite a low level operation.  A suggestion might be:

> Here we model abstractly how one interacts with files assuming a single thread that reads and writes, given a pointer into the file.  The pointer can be moved to an arbitrary position, or it can read or write data linearly from its current position, which simultaneously advances the pointer.  We can modify the length of the file and if the pointer would fall outside of the bounds after truncation, it is moved to the end of the file.